### ISSUE TITLE
Nginx Reverse Proxy Fails Due to CSP

### ISSUE THREAD

**Issue by OoLunar** [grafana/grafana#38468]
**What happened**:
![image](https://user-images.githubusercontent.com/46751150/130519668-03ea9b78-b030-43e0-aece-6ea0313ca82c.png)
Firefox:
> Uncaught EvalError: call to eval() blocked by CSP    core.js:4:29
> Content Security Policy: The page’s settings blocked the loading of a resource at eval (“default-src”). angular.js:1292:6
> Content Security Policy: The page’s settings blocked the loading of a resource at eval (“default-src”).     core.js:4:23

Chromium:
> Uncaught EvalError: Refused to evaluate a string as JavaScript because 'unsafe-eval' is not an allowed source of script in the following Content Security Policy directive: "default-src 'self' http: https: data: blob: 'unsafe-inline'".    core.js:4 

**What you expected to happen**:
![image](https://user-images.githubusercontent.com/46751150/130520297-72b654d0-704e-4172-8995-2f5fbc7676be.png)

**How to reproduce it (as minimally and precisely as possible)**:
Proxy port 3000 through Nginx using the Nginx guide found [here](https://grafana.com/tutorials/run-grafana-behind-a-proxy/) for instructions

**Anything else we need to know?**:
Nginx config (domain name redacted):
```nginx
map $http_upgrade $connection_upgrade {
  default upgrade;
  '' close;
}

server {
    listen 443 ssl http2;
    server_name grafana.example.com;

    # SSL, wildcard certificate
    ssl_certificate /etc/letsencrypt/live/example.com/fullchain.pem;
    ssl_certificate_key /etc/letsencrypt/live/example.com/privkey.pem;

    # logging
    error_log /var/log/nginx/grafana.example.com.error.log warn;

    location / {
        proxy_pass http://grafana:3000;
    }

    # Proxy Grafana Live WebSocket connections.
    location /api/live {
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "Upgrade";
        proxy_set_header Host $http_host;
        proxy_pass http://grafana:3000/;
    }
}

# HTTP redirect
server {
    listen 80;
    server_name grafana.example.com;
    return 301 https://grafana.example.com$request_uri;
}
```

`grafana.ini`: https://gist.github.com/OoLunar/6cf59ab2e2b69277475b4213d3d73cc3
Grafana output:
```log
[root@phoenixnetwork grafana]# docker-compose up -V
Recreating grafana_grafana_1 ... done
Attaching to grafana_grafana_1
✔ Downloaded grafana-clock-panel v1.0.1 zip successfully

Please restart Grafana after installing plugins. Refer to Grafana documentation for instructions if necessary.

✔ Downloaded grafana-simple-json-datasource v1.3.5 zip successfully

Please restart Grafana after installing plugins. Refer to Grafana documentation for instructions if necessary.

t=2021-08-23T21:13:45+0000 lvl=info msg="Starting Grafana" logger=server version=8.1.2 commit=103f8fa094 branch=HEAD compiled=2021-08-19T18:19:17+0000
t=2021-08-23T21:13:45+0000 lvl=info msg="Config loaded from" logger=settings file=/usr/share/grafana/conf/defaults.ini
t=2021-08-23T21:13:45+0000 lvl=info msg="Config loaded from" logger=settings file=/etc/grafana/grafana.ini
t=2021-08-23T21:13:45+0000 lvl=info msg="Config overridden from command line" logger=settings arg="default.paths.data=/var/lib/grafana"
t=2021-08-23T21:13:45+0000 lvl=info msg="Config overridden from command line" logger=settings arg="default.paths.logs=/var/log/grafana"
t=2021-08-23T21:13:45+0000 lvl=info msg="Config overridden from command line" logger=settings arg="default.paths.plugins=/var/lib/grafana/plugins"
t=2021-08-23T21:13:45+0000 lvl=info msg="Config overridden from command line" logger=settings arg="default.paths.provisioning=/etc/grafana/provisioning"
t=2021-08-23T21:13:45+0000 lvl=info msg="Config overridden from command line" logger=settings arg="default.log.mode=console"
t=2021-08-23T21:13:45+0000 lvl=info msg="Config overridden from Environment variable" logger=settings var="GF_PATHS_DATA=/var/lib/grafana"
t=2021-08-23T21:13:45+0000 lvl=info msg="Config overridden from Environment variable" logger=settings var="GF_PATHS_LOGS=/var/log/grafana"
t=2021-08-23T21:13:45+0000 lvl=info msg="Config overridden from Environment variable" logger=settings var="GF_PATHS_PLUGINS=/var/lib/grafana/plugins"
t=2021-08-23T21:13:45+0000 lvl=info msg="Config overridden from Environment variable" logger=settings var="GF_PATHS_PROVISIONING=/etc/grafana/provisioning"
t=2021-08-23T21:13:45+0000 lvl=info msg="Path Home" logger=settings path=/usr/share/grafana
t=2021-08-23T21:13:45+0000 lvl=info msg="Path Data" logger=settings path=/var/lib/grafana
t=2021-08-23T21:13:45+0000 lvl=info msg="Path Logs" logger=settings path=/var/log/grafana
t=2021-08-23T21:13:45+0000 lvl=info msg="Path Plugins" logger=settings path=/var/lib/grafana/plugins
t=2021-08-23T21:13:45+0000 lvl=info msg="Path Provisioning" logger=settings path=/etc/grafana/provisioning
t=2021-08-23T21:13:45+0000 lvl=info msg="App mode production" logger=settings
t=2021-08-23T21:13:45+0000 lvl=info msg="Connecting to DB" logger=sqlstore dbtype=sqlite3
t=2021-08-23T21:13:45+0000 lvl=info msg="Starting DB migrations" logger=migrator
t=2021-08-23T21:13:45+0000 lvl=info msg="migrations completed" logger=migrator performed=0 skipped=330 duration=1.358559ms
t=2021-08-23T21:13:45+0000 lvl=info msg="Starting plugin search" logger=plugins
t=2021-08-23T21:13:45+0000 lvl=info msg="Registering plugin" logger=plugins id=input
t=2021-08-23T21:13:45+0000 lvl=warn msg="Some plugin scanning errors were found" logger=plugins errors="plugin 'grafana-clock-panel' is unsigned, plugin 'grafana-simple-json-datasource' is unsigned"
t=2021-08-23T21:13:45+0000 lvl=info msg="Live Push Gateway initialization" logger=live.push_http
t=2021-08-23T21:13:45+0000 lvl=info msg="HTTP Server Listen" logger=http.server address=[::]:3000 protocol=http subUrl= socket=
t=2021-08-23T21:13:57+0000 lvl=info msg="Request Completed" logger=context userId=0 orgId=0 uname= method=GET path=/ status=302 remote_addr=47.4.110.2 time_ms=0 size=29 referer=
t=2021-08-23T21:14:15+0000 lvl=info msg="Request Completed" logger=context userId=0 orgId=0 uname= method=GET path=/ status=302 remote_addr=47.4.110.2 time_ms=0 size=29 referer=
```
Docker Compose File:
```yml
version: "3.9"

services:
  grafana:
    image: grafana/grafana:latest
    restart: unless-stopped
    networks:
      - influxdb
      - web_services
    volumes:
      - "./etc_grafana:/etc/grafana"
      - "./var_lib_grafana:/var/lib/grafana"
    ports:
      - 3000:3000
    environment:
      GF_INSTALL_PLUGINS: "grafana-clock-panel 1.0.1,grafana-simple-json-datasource 1.3.5"

networks:
  influxdb:
    external: true
  web_services:
    external: true
```

**Environment**:
- Grafana version: Version 8.1.2 (commit: 103f8fa094, branch: HEAD)
- Data source type & version:
- OS Grafana is installed on: Docker Alpine image
- User OS & Browser: Void Linux, Mozilla Firefox 91.0.1 and Chromium 91.0.4472.114
- Grafana plugins: grafana-clock-panel 1.0.1, grafana-simple-json-datasource 1.3.5
- Others: Cloudflare proxy is disabled for the `grafana` domain, Cloudflare is put in development mode, Cloudflare should not be affecting headers in any known way possible.

Yes, it works fine when I connect directly. It only fails during the Nginx Proxy. This leads me to think that the [docs](https://grafana.com/tutorials/run-grafana-behind-a-proxy/) I followed were either outdated or incorrect.

**Comment by hugohaggmark** [comment]
Passing this one over to @grafana/backend-platform

**Comment by marefr** [comment]
@OoLunar are you sure that you have configured the CSP setting correctly? Maybe try using triple quotes as in the default template, i.e. `content_security_policy_template = """script-src 'self' 'unsafe-eval' 'unsafe-inline';"""`. Maybe try changing `root_url = https://grafana.example.com`?

CSP is not enabled by default, you are using a custom CSP template and CSP is not mentioned in the tutorial you referenced. If Grafana works as expected without CSP enabled I'm not sure we can help you since it seems like a configuration mistake from your side either in Grafana or in your nginx proxy and it's very hard for us to investigate that further. If you turn off CSP in Grafana does it for example work as expected thru the nginx proxy? If you could put together steps re-producing a problem involving CSP and Grafana without nginx you can maybe open a new issue for that and we can look into it further, otherwise I would recommend you 
asking for help on our [community site](https://community.grafana.com/). Github are mainly for feature requests and bug reports.

If I've misunderstood something, please let us know and we'll consider reopening this.

**Comment by OoLunar** [comment]
Thank you for your helpful response. Unfortunately, this doesn't fix the issue.

> are you sure that you have configured the CSP setting correctly?

We originally didn't touch the CSP policy when this error occurred. Even with the `content_security_policy` setting set to `true` or `false`, the same error kept happening. 

> Maybe try using triple quotes as in the default template, i.e. content_security_policy_template = """script-src 'self' 'unsafe-eval' 'unsafe-inline';""". Maybe try changing root_url = https://grafana.example.com?

I've done both, yet no difference.

>  If Grafana works as expected without CSP enabled I'm not sure we can help you since it seems like a configuration mistake from your side either in Grafana or in your nginx proxy and it's very hard for us to investigate that further.

At first, I too thought it was a configuration error. Then I created a fresh, new instance and the error kept popping up. I didn't touch anything related to the CSP policy, which is why I thought it was an application bug and opened up this issue. As I said previously, this works just fine when accessing it directly, however when accessing it through Nginx, the CSP error pops up. 

I've made sure to include all my configs in an attempt to make debugging easier, however with there being so much to configure, I'm sure debugging is no easy task. I will do everything in my power to help, and will continue to try new things on a different instance in order to get things working. Thank you for your support!

**Comment by OoLunar** [comment]
Everything works as expected on a new Nginx instance with no default configuration. This is definitely an issue on my end.

**Comment by OoLunar** [comment]
Going through my `nginx.conf`, turns out that I this was present:
```nginx
add_header Content-Security-Policy "default-src 'self' http: https: data: blob: 'unsafe-inline'" always;
```
Why this took precedence over the proxy headers is beyond me. Note for self, and others: **Check your subdomain config and Nginx config when debugging.**

I apologize for troubling y'all.

**Comment by marefr** [comment]
No worries. Glad you figured it out and thanks for letting us and others know
